contracts/contracts/lib/AssetHelper.sol

```diff
@@ -192,15 +197,29 @@ library SupplySourceHelper {
     }
 
     /// @notice Gets the balance in our treasury.
-    function balance(SupplySource memory supplySource) internal view returns (uint256 ret) {
-        if (supplySource.kind == SupplyKind.Native) {
+    function balance(Asset memory genericToken) internal view returns (uint256 ret) {
+        if (genericToken.kind == AssetKind.Native) {
             ret = address(this).balance;
-        } else if (supplySource.kind == SupplyKind.ERC20) {
-            ret = IERC20(supplySource.tokenAddress).balanceOf(address(this));
+        } else if (genericToken.kind == AssetKind.ERC20) {
+            ret = IERC20(genericToken.tokenAddress).balanceOf(address(this));
         }
     }
 
-    function native() internal pure returns (SupplySource memory) {
-        return SupplySource({kind: SupplyKind.Native, tokenAddress: address(0)});
+    function native() internal pure returns (Asset memory) {
+        return Asset({kind: AssetKind.Native, tokenAddress: address(0)});
+    }
+
+    function isNative(Asset memory self) internal pure returns(bool) {
+        return self.kind == AssetKind.Native;
+    }
+
+    function increaseAllowance(Asset memory self, address spender, uint256 amount) internal {
+        if (self.kind == AssetKind.ERC20) {
+            IERC20 token = IERC20(self.tokenAddress);
+            uint256 allowance = token.allowance(address(this), spender);
+            if (!token.approve(spender, allowance + amount)) {
+                revert IncreaseAllowanceFailed();
+            }
+        }
```

Comment:
I think this way might be more specific to handle `AssetKind` for ERC20. Like you mentioned below if there is a new variant, then the handling would be wrong.

Reply:
In this case it could never be wrong even if you add more type. Because it always deals with one specific type.

Reply:
understand where you are coming from. But say if we have another asset, NFT, then perhaps the current style can prevent accidently calling allowance on NFT? Actually I dont see much diff here because either way it will fail due to no allowance/permission. I will take note to check this out if we have any new type, but in the foreseeable short/long run, I dont think much will change.